### ISSUE TITLE
Dependencies: Refactor dependencies config into key-value object

### ISSUE THREAD

**Issue by fubuloubu** [ApeWorX/ape#398]
Currently the keys are like:

```yaml
dependencies:
  MyDependency: github-org/github-repo@1.2.3
```

Which is not very scaleable to other types of dependencies from other locations

Suggestion to break it out a bit:
```yaml
dependencies:
  - name: MyDependency
    github: github-org/github-repo
    version: 1.2.3
```

At first, only `github` and `version` keys are specified. Eventually relax the requirement of `version` key to allow alternatives like `branch`, and then eventually add other types of dependencies such as `gitlab`, `ipfs`, `ipns`, etc.

**Comment by unparalleled-js** [comment]
Talked elsewhere but the 'type' is not really needed, the key is the type. yaml.

**Comment by fubuloubu** [comment]
> Talked elsewhere but the 'type' is not really needed, the key is the type. yaml.

some "types" might not be excplicit, like github vs. gitlab, IPFS links, etc.

although I guess write up more what you mean in the resulting yaml?

**Comment by unparalleled-js** [comment]
> although I guess write up more what you mean in the resulting yaml?

I was only referring to how we use the schema to know the type, thinking about our convo regarding the pydantic config utility. We should be able to migrate to that easily if needed! I have PR in my local fork that resolves this issue (will open after #420)
